hathor/p2p/sync_checkpoints.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from itertools import chain
+from typing import TYPE_CHECKING, Dict, List, NamedTuple, Optional
+
+from structlog import get_logger
+from twisted.internet.task import LoopingCall
+
+from hathor.checkpoint import Checkpoint
+from hathor.transaction.block import Block
+from hathor.transaction.storage.exceptions import TransactionDoesNotExist
+from hathor.util import Reactor
+
+if TYPE_CHECKING:
+    from hathor.manager import HathorManager  # noqa: F401
+    from hathor.p2p.node_sync_v2 import NodeBlockSync  # noqa: F401
+
+logger = get_logger()
+
+
+class _SyncInterval(NamedTuple):
+    start_hash: bytes
+    start_height: int
+    end_hash: bytes
+    end_height: int
+
+
+class SyncCheckpoint:
+    """This is the central manager of the sync between checkpoints among all peers.
+    """
+
+    # Looping call interval.
+    LC_INTERVAL: int = 5
+
+    def __init__(self, manager: 'HathorManager'):
+        # All peers that have all the checkpoints to download
+        self.peers_to_request: List['NodeBlockSync'] = []
+
+        # All peers that we are connected but don't have all the checkpoints
+        self.incomplete_peers: List['NodeBlockSync'] = []
```

Comment:
That works too, the slight penalty is having to iterate over that dict to find a suitable peer when one is needed. If we maintain a `peers_to_request` they're readily available. However this shouldn't be a problem, the order of magnitude of connected peers should never be large enough that this would make a difference.

Reply:
Moved to another PR.